### DISCUSSION TITLE
Problems with step time of Windfreak Microwave generator

### DISCUSSION THREAD

**Discussion by JohannesWesseler** [Ulm-IQO/qudi-core#68] [Q&A] [answered]
Hi,
I am using the windfreak SynthNV Pro for microwave generation and have a question regarding the interplay between qudi, the windfreak and the NI DAQ card. I slightly modified the SynthHD Pro hardware file, but only to adapt for different naming of some commands, e.g. the "w" command of the SynthHD Pro is called "y" for the SynthNV Pro but apart from that is identical, so the two devices should also mostly behave the same.

I am using the mode "EQUIDISTANT_SWEEP". My windfreak is connected via usb to my pc and I trigger the windfreak using a digital output of a NI 782253-01 (USB 6343).

### **My problem is, that the microwave frequencies in qudi and the microwave frequencies that the windfreak sends out do not seem to be synced correctly.**

The symmetry axis/zfs of the ODMR axis initially seems to be at smaller microwave frequencies than 2.87 GHz. When I change the interval of frequencies that I use, the symmetry axis of the ODMR curve moves closer to 2.87 GHz.

Largest interval:

![20230914-1612-15_20230914-1608-54_wo_ist_die_mitte_bildausschnitt_max_1W_laser_no_B_field_ODMR_signal_ODMR_ai0_range0](https://github.com/Ulm-IQO/qudi-core/assets/142618752/ff7efa18-0497-462b-82d9-be7f28666b2c)

Intermediate interval:

![20230914-1608-54_wo_ist_die_mitte_bildausschnitt_gross_1W_laser_no_B_field_ODMR_ai0_range0](https://github.com/Ulm-IQO/qudi-core/assets/142618752/fb4ac500-745c-4f4c-ae37-ad5b691e356d)

Smallest interval:

![20230914-1605-55_wo_ist_die_sym-achse_1W_laser_no_B_field_ODMR_ai0_range0](https://github.com/Ulm-IQO/qudi-core/assets/142618752/01a28979-c5f0-4432-8e41-9d0e6d26437c)

A similar problem occurs when I leave the frequency interval fixed, but change the data rate in the ODMR settings dialog box.

This is the result when using a data rate of 200Hz:

![20230915-1120-29_peak_at_2 83_data_rate_200Hz_ODMR_ai0_range0](https://github.com/Ulm-IQO/qudi-core/assets/142618752/9cc1ff37-2e76-4c21-aefa-5f487ed0fcca)

This for 100 Hz:

![20230915-1123-55_dip_bei_2 85_datarate100Hz_ODMR_ai0_range0](https://github.com/Ulm-IQO/qudi-core/assets/142618752/3b5839fd-32f9-4f76-9086-c702abd48a6d)

And this for 50 Hz:

![20230915-1125-32_dip_bei_2 87_datarate50Hz_ODMR_ai0_range0](https://github.com/Ulm-IQO/qudi-core/assets/142618752/cbcb7e1a-ed4a-426f-8773-adcff1207481)

The curves clearly move to the right with increasing data rate.

Has someone experienced this before and knows a solution? I think it might be connected to the values of step times / sweep rates / triggering rates.

**Comment by NicolasStaudenmaier** [comment]
Hi @JohannesWes, 
thanks for reaching out with your problem. I assume that this might be related to the windfreak SynthNV Pro. I have tried to reproduce what you observe with the same NIDAQ (USB 6343), but with a Rohde & Schwarz SMIQ. I have tried both modes, JUMP_LIST (default option, that I always use) and EQUIDISTANT_SWEEP, both with different frequency ranges and different sample rate. I could not observe any difference in the measured data, in particular of the determined resonance. 

Hence, the ODMR toolchain must be working well. However, I use detector with digital counts, but I heavily doubt that that could be the reason. I will try to find and lend a windfreak microwave generator I to verify better. For the meantime, what happens when the expected signal dip is at the edge of the frequency range? Will the dip be split left and right? If possible you can also try to play a negative sweep, i.e. starting from a higher frequency and going to lower ones. How does it look in that case?

**Reply by JohannesWesseler** [reply]
Thank you for yours answer. I think the JUMP_LIST mode is not properly implemented in qudi yet for the windfreak, based on what I see in the hardware file. However I also think, that the problem is not on the qudi side but more with the timing between windfreak and the daq card. I thought maybe the people that implemented the windfreak synthHD hardware file had experienced similar errors, but if this is not the case, we will see from what else these problems are caused. I do not think the problems are due to us using the SynthNV instead of the SynthHD, the devices should function almost identically.

For the expected signal dip at the left edge of the frequency range, the dip does nto move when I change the left interval boundary:

![20230918-1831-07_linked_ende_2 78_ODMR_ai0_range0](https://github.com/Ulm-IQO/qudi-core/assets/142618752/49f806bc-b0c0-479a-9b33-6be78be03554)
![20230918-1832-52_linked_ende_2 86_ODMR_ai0_range0](https://github.com/Ulm-IQO/qudi-core/assets/142618752/5da87810-662f-4634-b601-5ee9a5049751)
![20230918-1833-31_linked_ende_2 87_ODMR_ai0_range0](https://github.com/Ulm-IQO/qudi-core/assets/142618752/3b4f9de6-db58-480e-9e81-56e19c5d0b47)


For the signal dip at the right edge of the interval, the middle of the peaks moves closer to 2.87 GHz and then when we go even closer to the edge, the curve flattens out, very weird.
![20230918-1822-17_rechtes_ende_2 95_ODMR_ai0_range0](https://github.com/Ulm-IQO/qudi-core/assets/142618752/f8d3cc3b-75fc-47d9-88c1-c495d9644053)
![20230918-1824-06_rechtes_ende_2 88_ODMR_ai0_range0](https://github.com/Ulm-IQO/qudi-core/assets/142618752/b97f44b7-86ee-4924-9771-94d953b5b1b5)
![20230918-1828-39_rechtes_ende_2 865_ODMR_ai0_range0](https://github.com/Ulm-IQO/qudi-core/assets/142618752/40ccd2df-4032-4ed6-816c-c79434ccfff7)

For  a negative sweep, we get just noise. Here I think due to the timing errors, the frequency difference between what qudi expects for the MW frequency and the current frequency of the WF generator are different for each line, leading to the noise.

![20230918-1834-59_negative_frequency_noise_ODMR_ai0_range0](https://github.com/Ulm-IQO/qudi-core/assets/142618752/76c14298-f97b-4f56-9d00-1b26508ff60f)

**Comment by JohannesWesseler** [comment, marked as answer]
This is mainly solved now. We didn't consider that the windfreak is low active, so when the nidaq idle state was on high, the windfreak already starts sweeping before the sweep is supposed to start. Thus the line is shifted to the left. This was not considered in the windfreak synthHD pro qudi hardware file, so I would say this hardware file should be seen as faulty.

**Comment by NicolasStaudenmaier** [comment]
Great that you found the problem. If you have fixed it I would like to encourage you to create a pull request in the qudi-iqo-modules repository with your bug fix, such that others can profit as well and don't have to run into the same error. If not please write at least an issue to keep track of current bugs. Thank you!

**Reply by JohannesWesseler** [reply]
Hi. The thing is, that because the windfreak is low-active, for my fix I also modified the ni_x_series_finite_sampling_input.py file, such that the idle_state is ni.constants.Level.HIGH instead of ni.constants.Level.LOW. Thus only changing the windfreak hardware file is not sufficient and I think an issue is the better choice.
